### PR TITLE
[CFDX-785] add all context to perread

### PR DESCRIPTION
**Pull request by Javkhaa**
This PR adds methylation call string like Bismarks XM field into perRead command.

Example output:
```
A00949:398:H35KNDSXC:1:2101:19072:9627  chr1    88264   10.204082       49      XM:Z:h.hx.....Hh.x...hx.....h.h...x..x..h.h....hh.hhhh.Z...h..hh....h.....x....h.h......h.h.h.hhh..hxZ.h..h.x........hhhh.hh.xZ.Z......
A00949:398:H35KNDSXC:1:2101:19072:9627  chr1    88281   8.163265        49      XM:Z:x.....h.h...x..x..h.h....hh.hhhh.Z...h..hh....h.....x....h.hx.....h.h.h.hhh..hxZ.h..h.x........hhhh.hh.xZ.Z......x..h..h..hh...x..
```
and the corresponding Bismark BAM has this
```
A00949:398:H35KNDSXC:1:2101:19072:9627_1:N:0:TGATAGGC+ACTCAGAC  99      chr1    88265   22      130M    =       88282   147     TTTTTGGGACTATAGGTTTGTGTTTTTATTTGTTGTTTAATATTTTTTTTCGAATATTTATGGTTAAAATTGAGTTTATGATATTTTTTTTTTATTCGTTTTTTTGTGGATTTTTTATTTTCGCGAAGGA      FFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFF:FFFFFFFFFFFFFFF:FF,FFFFFFFFFFFFFFFFFFFFF      NM:i:45 MD:Z:0C1C0C6C1C3C0C5C1C3C2C2C1C4C0C1C0C0C0C5C2C0C4C5C4C1C0C5C1C1C1C0C0C2C0C2C2C1C8C0C0C0C1C0C1C9        XM:Z:h.hx.....Hh.x...hx.....h.h...x..x..h.h....hh.hhhh.Z...h..hh....h.....x....h.h......h.h.h.hhh..hxZ.h..h.x........hhhh.hh.xZ.Z...... XR:Z:CT XG:Z:CT
A00949:398:H35KNDSXC:1:2101:19072:9627_1:N:0:TGATAGGC+ACTCAGAC  147     chr1    88282   22      130M    =       88265   -147    TTGTGTTTTTATTTGTTGTTTAATATTTTTTTTCGAATATTTATGGTTAAAATTGAGTTTTTGATATTTTTTTTTTATTCGTTTTTTTGTGGATTTTTTATTTTCGCGAAGGATAGTTTTATTTTTTTAG      FFFFFFFF,FFFFFFFFFF::FFFFFF::FFFFFFFFFFFFFFFFFFF:FFFFFFFF:FFFFFFFF,FF::FFFFFFFF,FFF:F:FFFFFFFFFFFFFFFFFF:F:FFFFFFFFFFFFFFF:FFFFFFF      NM:i:45 MD:Z:0C5C1C3C2C2C1C4C0C1C0C0C0C5C2C0C4C5C4C1C0C5C1C1C1C0C0C2C0C2C2C1C8C0C0C0C1C0C1C9C2C2C2C0C3C2        XM:Z:x.....h.h...x..x..h.h....hh.hhhh.Z...h..hh....h.....x....h.hx.....h.h.h.hhh..hxZ.h..h.x........hhhh.hh.xZ.Z......x..h..h..hh...x.. XR:Z:GA XG:Z:CT
```

So far they are identical, full BAM matching is in progress.

### CODE REVIEW

File: common.c [hunk 1]
```diff
@@ -81,6 +81,12 @@ inline int isCHH(char *seq, int pos, int seqlen) {
     return 0;
 }
 
+inline int isUnknownC(char *seq, int pos, int seqlen) {
+    if(pos >= seqlen) return 0;
```
Comment (robertsugar) [inline common.c:85]: Q: do you know what happens with indels? (for example if a C is deleted or a new CG is inserted?)

**Review by robertsugar — CHANGES_REQUESTED** [review]
Wow! Total flashback to C :)

One question about indels, and a request to add some tests for some edge cases.

Q1: does this code work with Bismark BAM? we could use that to see if they are identical for random 1 million reads or so
Q2: add the test case for a small bwa-meth + bismark bam to see the similarities / differences on the same fastq
Q3. maybe some contrived examples of edge cases

Note: I would do some of these that are easy, for potential chunky bits, just add an engineering ticket.

File: common.c (hunk 1 above)
Comment (Javkhaa) [reply common.c:85]: so any insertion becomes unknown same as how bismark marks them, 
for deletion it will not consider the deleted C but rather consider the next base after the deleted sequence from reference.

**Comment by Javkhaa** [comment]
@robertsugar alright i have fixed all the bugs and issues haha 
I also added a test BAM file that includes different types of edge cases. added some description to the test folder.

File: common.c (hunk 1 above)
Comment (Javkhaa) [reply common.c:85]: OK now it matches perfectly.

**Review by robertsugar — APPROVED** [review]
Looks good.

one comment (on the workflow, not on this PR - would be great to have this zipped with bgzip instead of gzip, and named xm.gz instead of csv.gz -> this will allow us to separate these files by extension, and to easily tabix it for region extraction)

**Comment by Javkhaa** [comment]
> Looks good.
> 
> one comment (on the workflow, not on this PR - would be great to have this zipped with bgzip instead of gzip, and named xm.gz instead of csv.gz -> this will allow us to separate these files by extension, and to easily tabix it for region extraction)

Sounds good, oh yeah forgot about the nextflow pipeline let me add it to bioinformatics repo
